File: internal/worktree/git/topo_paths.go
package git

import (
	"crypto/sha256"
	"encoding/hex"
	"os"
	"path/filepath"
)

const topoDuctorDir = ".topoDuctor"

func topoDuctorRoot() (string, error) {
	h, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(h, topoDuctorDir), nil
}

// projectSegmentName carpeta bajo projects/ a partir de la ruta del repo (única por clone).
func projectSegmentName(repoTop string) string {
	base := filepath.Base(repoTop)
	slug := SanitizeWorktreeLabel(base)
	if slug == "" {
		slug = "repo"
	}
	sum := sha256.Sum256([]byte(filepath.Clean(repoTop)))
	short := hex.EncodeToString(sum[:4])
	return slug + "-" + short
}

// checkoutPathForNewWorktree devuelve ~/.topoDuctor/projects/<segmento>/worktree/<wtSlug>.
func checkoutPathForNewWorktree(repoTop, wtSlug string) (string, error) {
	root, err := topoDuctorRoot()
	if err != nil {
		return "", err
	}
	seg := projectSegmentName(repoTop)
	return filepath.Join(root, "projects", seg, "worktree", wtSlug), nil
}

File: ink-app/src/git/topoPaths.ts
import crypto from "node:crypto";
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { sanitizeWorktreeLabel } from "./sanitize.js";

const topoDuctorDir = ".topoDuctor";

async function topoDuctorRoot(): Promise<string> {
  return path.join(os.homedir(), topoDuctorDir);
}

export function projectSegmentName(repoTop: string): string {
  let base = path.basename(repoTop);
  let slug = sanitizeWorktreeLabel(base);
  if (!slug) {
    slug = "repo";
  }
  const sum = crypto.createHash("sha256").update(path.normalize(repoTop)).digest();
  const short = sum.subarray(0, 4).toString("hex");
  return `${slug}-${short}`;
}

/** ~/.topoDuctor/projects/<segment>/worktree/<wtSlug> */
export async function checkoutPathForNewWorktree(
  repoTop: string,
  wtSlug: string
): Promise<string> {
  const root = await topoDuctorRoot();
  const seg = projectSegmentName(repoTop);
  return path.join(root, "projects", seg, "worktree", wtSlug);
}

export async function ensureDirForFile(filePath: string): Promise<void> {
  await fs.mkdir(path.dirname(filePath), { recursive: true, mode: 0o755 });
}
